### ISSUE TITLE
Documentation improvements

### ISSUE THREAD

**Issue by michaelbromley** [vendure-ecommerce/vendure#2299]
## Background

Vendure is dedicated to providing an outstanding developer experience. A major part of that is our documentation at https://docs.vendure.io. The docs have a solid foundation of correctness (all API docs are generated from source in a readable, linked way) and coverage in the form of many pages of hand-written guides.

However, in the spirit of continual improvement, I am launching an effort to further improve the documentation with the following goals (in order of priority):

- Make it easier to find the information you are looking for
- Fill in missing docs, especially step-by-step guides for common tasks
- Improve the overall look of the docs

## Improving information architecture

Right now we have a single navbar on the left with a tree structure. There is a visual break between the hand-written guides and the generated API docs: 
![image](https://github.com/vendure-ecommerce/vendure/assets/6275952/47c60150-5acc-49ee-9c88-32acab330f52)

I think we can improve the information architecture to something along the lines of the new React docs: https://react.dev/learn

They make the top-level split between "learn" (guides) and "reference" (API docs).

Within "learn" they have a "get started" and "learn react" sub-sections, and then these break down into the major conceptual areas: "describing the ui", "adding interactivity" etc. Each topic is a very comprehensive, tutorial-like article:

![image](https://github.com/vendure-ecommerce/vendure/assets/6275952/380a650a-7330-4cc0-8bff-04f0044479ad)

In our case, the "learn vendure" topic areas could be something like:

* Writing a plugin
* Extending the Admin UI
* Deployment
* Building a storefront

The "administrator guide" which contains instructions for using the Admin UI could be moved to a different section, e.g. "operator's manual".

### Task-focused

We should move towards a more task-focused approach. There are common tasks that devs want to complete with Vendure - we need to anticipate these and provide a complete, step-by-step guide to doing it, including copy-pastable code snippets.

More complex tasks that require many related parts in different files can be complemented with an actual working plugin in the dev-server/documentation-examples directory.

### Improved search 

Feedback from a user:
>I’d like to make a case for better search. I think the documentation has most of the information required to customize Vendure. 
> 
> However, the documentation is laid out based on code and objects which means that I have to know Vendure and specific language to use the documentation which is counter intuitive.

We can improve search using keywords, and perhaps a full-on search engine rather than the client-side index search we currently have.

**Comment by michaelbromley** [comment]
> I can only agree to the things said above. One thing I would like to add is some kind of "Nest.js in the context of Vendure". It took me a long time to understand these decorators and the architecture with resolvers, modules, providers and services. Such a video or written guide would help immensely in the beginning when one is writing the first plugins.

**Comment by msj121** [comment]
There have been a few times where I have searched docs and had been pointed to somewhere else in docs. I feel like sometimes it can be tricky to find all the locations, even just a "look here for more" link where docs might overlap?

For example:  I followed https://docs.vendure.io/typescript-api/core-plugins/payments-plugin/braintree-plugin/ and although it mentioned database update/sync this was something I thought happened automatically and not something I needed to configure. Pointing to here https://docs.vendure.io/developer-guide/migrations/#synchronize-vs-migrate would have cleared it up? Just food for thought.

**Comment by michaelbromley** [comment]
Other guides that need writing:

- How to set up codegen
- GraphQL for Vendure developers

**Comment by Caryntjen** [comment]
What I struggled with was the migration (I generated duplicates, without running the migration and restarting the app), this of course is covered in other guides. But I think some best practices on migrations.

**Comment by michaelbromley** [comment]
Preview of ongoing work: https://new-docs--vendure-docs-beta.netlify.app/guides/getting-started/installation

**Comment by tianyingchun** [comment]
beautyfull UI interface!!!!

**Comment by NoahPerez** [comment]
If you can add in the marketplace docs, fundamentalmente parts needed to build a marketplace. I think it would would be the only one in the world open source fulling documented with cutting edge tech.

**Comment by zehawki** [comment]
Running list of some findings:
1. Admin UI > "Create new channel" > fields "Code" & "Channel token" need some explanation
2. Mention in https://docs.vendure.io/developer-guide/configuration/ that `VendureConfig` object is in `src/vendure-config.ts`

**Comment by mschipperheyn** [comment]
I have recently started working with Vendure. As impressed as I am with the toolbox, there are a few areas where I really went down rabbit holes as I started building a custom plugin that required fields from other custom plugins. I would suggest adding documentation on how to ensure you can access customFields and extensions on e.g. OrderLine or Customer inside a custom plugin resolver you are building, when those extensions were made by other plugins you are loading.

What I run into is that sometimes properties are loaded by services offered by those third party plugins or you are using the @vendure/core services. But in both cases, I'm struggling to get access to customFields created by other plugins that I know are there, but are not being loaded. Since you you are already "behind the resolver in the service layer", you can no longer rely on any kind of resolver based resolution to get those. Anyways, would love to get a howto on that.

**Comment by mschipperheyn** [comment]
Hi @michaelbromley let me try to give you that example in a summary way

So, I'm building a plugin that uses two existing plugins, one is [@pinelab customer-managed-groups](https://github.com/Pinelab-studio/pinelab-vendure-plugins/tree/main/packages/vendure-plugin-customer-managed-groups). This plugin allows you to assign users that you consider as part of your group, e.g. family members. Another plugin we use allows you to assign that group member to an orderline, so you can buy products for your child.

```
import {
  Customer,
  LanguageCode,
  PluginCommonModule,
  VendurePlugin,
  Allow,
  Transaction,
} from "@vendure/core"
import gql from "graphql-tag"
import { Args, Mutation, Resolver } from "@nestjs/graphql"
import { Ctx, RequestContext, OrderService } from "@vendure/core"
import { Permission, OrderLine } from "@vendure/common/lib/generated-types"
import { isFunctionTypeNode } from "typescript"

const assignCustomerSchemaExtension = gql`
  extend type Mutation {
    assignCustomerToOrderLine(orderLineId: ID!, customerId: ID!): OrderLine!
  }
`

@Resolver("OrderLine")
export class OrderLineAssignmentResolver {
  constructor(private orderService: OrderService) {
    console.log("OrderLineAssignmentResolver constructor called")
  }

  @Transaction()
  @Mutation()
  @Allow(Permission.UpdateCatalog)
  async assignCustomerToOrderLine(
    @Ctx() ctx: RequestContext,
    @Args() args: any
  ) {
    const order = await this.orderService.findOneByOrderLineId(
      ctx,
      args.orderLineId
    )
    const orderId = order!.id
    console.log("assignCustomerToOrderLine called with args: ", args)
    // error checking
    return this.orderService.adjustOrderLine(
      ctx,
      orderId,
      args.orderLineId,
      1,
      { assignedCustomer: { id: args.customerId } }
    )
  }
}

@VendurePlugin({
  imports: [PluginCommonModule],
  shopApiExtensions: {
    schema: assignCustomerSchemaExtension,
    resolvers: [OrderLineAssignmentResolver],
  },
  configuration: (config) => {
    config.customFields.OrderLine.push({
      name: "assignedCustomer",
      type: "relation",
      entity: Customer,
      eager: false,
      defaultValue: false,
      nullable: true,
      label: [{ languageCode: LanguageCode.en, value: "Assigned Member" }],
    })
    return config
  },
  compatibility: "^2.0.0",
})
export class OrderLineAssignmentPlugin {}

```
One feature is that the plugin adds a customField to an OrderLine.

So, on my side, I'm building a plugin that ensures that you can't select any products that were already bought for user X or family member Y. So, I use pinelabs order retrieval for family members and try to verify that assignedCustomer. It doesn't exist on that (I guess retrieval is not using *) and also when I use the @vendure/core order manager, I cannot access the assignedCustomer on the retrieved orderLine. 

So, I'm looking for how to get that assignedCustomer loaded up and have it recognized as a valid typescript property. That last part I think I'll be able to get through. The first part is trickier.

HTH

**Comment by michaelbromley** [comment]
Thanks for this detailed explanation, it is helpful!

So regarding the TS typings for custom fields, this is covered here: https://beta-docs.vendure.io/guides/developer-guide/custom-fields/#typescript-typings - let me know if that is helpful on that point.

Regarding the broader issues brought up, I think we need 2 sections added to the docs:

### 1. Working with relations

This section could be a sub-heading of [this section](https://beta-docs.vendure.io/guides/developer-guide/the-service-layer/#accessing-the-database) which explains how relations work, i.e.

- There's no build-time guarantee of whether relations will be joined at runtime
- Explain the use of [EntityHydrator](https://beta-docs.vendure.io/reference/typescript-api/data-access/entity-hydrator/) to ensure the required relations are joined.
- Many of the core service methods include a `relations` argument which allows you to specify which relations to load, e.g.

```ts
const order = await this.orderService.findOne(ctx, id, ['lines.customFields.assignedCustomer']);
```

### 2. Accessing custom fields via TypeScript code

A new section in the [custom fields guide](https://beta-docs.vendure.io/guides/developer-guide/custom-fields/) which demonstrates patterns for accessing custom fields, e.g.

```ts
const orderLine = await this.connection.getRepository(ctx, OrderLine).findOne({
  where: { id },
  relations: {
    customFields: { 
      assignedCustomer: true,
    }
  }
})
```

or with EntityHydrator:

```ts
const order = await this.orderService.findOne(ctx, id);
await this.entityHydrator.hydrate(ctx, order, { relations: ['lines.customFields.assignedCustomer'] });
```

**Comment by michaelbromley** [comment]
One to add the the Admin UI customization docs:

- https://github.com/vendure-ecommerce/vendure/discussions/2368

**Comment by michaelbromley** [comment]
Add section on all aspects of email handling, including how to toggle dev/prod mode in the email config:

```ts
  EmailPlugin.init({
        handlers: [
            // ...
        ],
        templatePath: path.join(rootDir, 'static/email/templates'),
        globalTemplateVars: {
            // ...
        },
        ...(DEV_MODE
            ? {
                  route: 'mailbox',
                  devMode: true,
                  outputPath: path.join(rootDir, 'static/email/output'),
              }
            : {
                  transport: {
                      type: 'smtp',
                      host: process.env.SMTP_HOST,
                      port: +process.env.SMTP_PORT,
                      //...
                  },
              }),
```

**Comment by mschipperheyn** [comment]
Another suggestion I have is to add a section to Error Handling that explains how to create your own Errors, in particular how to ensure that say a validation error gets converted to an ErrorResult if you use an ErrorResult union. My first impression is that it's prob not desirable to extend on Vendure's error result. I'm not even sure if it's possible to extend that `ErrorCode` enum. So, it's prob more of a nestjs style question but it's a scenario that's likely common if you're building your own plugins.

**Comment by tianyingchun** [comment]
yes, it should be expose on docs :)
```ts
  extend enum ErrorCode {
    CROSS_EMAIL_AUTHENTICATION
  }
  type CustomPluginError implements ErrorResult {
    errorCode: ErrorCode!
    message: String!
  }
```
```ts

export class CustomPluginError {
  // eslint-disable-next-line @typescript-eslint/naming-convention
  readonly __typename = 'CustomPluginError';

  constructor(
    private readonly message: string,
    private readonly errorCode: string
  ) {
    console.log('runhere');
  }
}
```
```ts
return new RequestOneTimeCodeError(
          `Email already used with "${foundStrategy}" authentication`,
          ErrorCode.CROSS_EMAIL_AUTHENTICATION
        );
```

**Comment by mschipperheyn** [comment]
From what I can tell from the issues (https://github.com/vendure-ecommerce/vendure/issues/437) and the repo there is actually a plugin for this

However, on my side, I do see the enum `ErrorCode` values being auto generated but not the error class being auto generated. I'm not so familiar with codegen and I get how it's supposed to work through `graphql-errors-plugin` but all of this is a bit auto-magical and hard to debug.

**Comment by mschipperheyn** [comment]
I finally got it. I have been dumb, but it is an easy one to trip over if you don't use you brain. You have use the `isGraphQlErrorResult` (or some other method) to determine if the error that is occurring in your resolver method is a an ErrorResult. In that case, you have to return it in stead of throwing it.

**Comment by mschipperheyn** [comment]
Documentation improvements: Add fields to existing types

Some aspects on adding custom fields to existing domain objects is not clear in the documentation:
Context: I want to add an avatar to the Administrator. 

1. If you add a custom field to an existing entity, that it gets added to the graphql api automatically (If you try to do it manually you run into a dead end because custom field type definitions are created in the code generation step)
2. How to add a custom field to the root of an existing entity instead of to the `customFields` object. The ProductVariant example makes clear that it's possible but not how to configure it at the entity level, since the documented way adds fields at the customFields level.
3. How to avoid the Error: "Vendure Entity you are extending" defined in resolvers, but not in schema.The custom fields get generated automatically, but I need an entity resolver to retrieve the Asset that is added as the avatar field to customFields. 

I added the avatar field as a custom field. Then to avoid ad 3, I tried to add something like 

```
extend type AdministratorCustomFields {
   avatar: Asset
}

extend type Administrator {
    customFields: AdministratorCustomFields
}
```

But since AdministratorCustomFields didn't exist yet, I ran into ad 1. 

```
extend type Administrator {
    customFields: {
        avatar: Asset
   }
}
```
is illegal
